apis/postgresql/v1alpha1/role_types.go

```diff
@@ -0,0 +1,121 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+// A RoleSpec defines the desired state of a Role.
+type RoleSpec struct {
+	xpv1.ResourceSpec `json:",inline"`
+	ForProvider       RoleParameters `json:"forProvider"`
+}
+
+// A RoleStatus represents the observed state of a Role.
+type RoleStatus struct {
+	xpv1.ResourceStatus `json:",inline"`
+	AtProvider          RoleObservation `json:"atProvider,omitempty"`
+}
+
+// RolePrivilege is the PostgreSQL identifier to add or remove a permission
+// on a role.
+// See https://www.postgresql.org/docs/current/sql-createrole.html for available privileges.
+// INHERIT;NOINHERIT; not implemented here as we don't currently implement role membership.
+type RolePrivilege struct {
+	// SuperUser grants SUPERUSER privilege when true.
+	// +optional
+	SuperUser *bool `json:"superUser,omitempty"`
+
+	// CreateDb grants CREATEDB when true, allowing the role to create databases.
+	// +optional
+	CreateDb *bool `json:"createDb,omitempty"`
+
+	// CreateRole grants CREATEROLE when true, allowing this role to create other roles.
+	// +optional
+	CreateRole *bool `json:"createRole,omitempty"`
+
+	// Login grants LOGIN when true, allowing the role to login to the server.
+	// +optional
+	Login *bool `json:"login,omitempty"`
+
+	// Inherit grants INHERIT when true, allowing the role to inherit permissions
+	// from other roles it is a member of.
+	// +optional
+	Inherit *bool `json:"inherit,omitempty"`
+
+	// Replication grants REPLICATION when true, allowing the role to connect in replication mode.
+	// +optional
+	Replication *bool `json:"replication,omitempty"`
+
+	// BypassRls grants BYPASSRLS when true, allowing the role to bypass row-level security policies.
+	// +optional
+	BypassRls *bool `json:"bypassRls,omitempty"`
+}
+
+// RoleParameters define the desired state of a PostgreSQL role instance.
+type RoleParameters struct {
+	// ConnectionLimit to be applied to the role.
+	// +kubebuilder:validation:Min=-1
+	// +optional
+	ConnectionLimit *int32 `json:"connectionLimit,omitempty"`
+
+	// Privileges to be granted.
+	// +optional
+	Privileges RolePrivilege `json:"privileges,omitempty"`
+
+	// PasswordSecretRef references the secret that contains the password used
+	// for this role. If no reference is given, a password will be auto-generated.
+	// +optional
+	PasswordSecretRef *xpv1.SecretKeySelector `json:"passwordSecretRef,omitempty"`
+}
+
+// A RoleObservation represents the observed state of a PostgreSQL role.
+type RoleObservation struct {
+	// PrivilegesAsClauses represents the applied privileges state, taking into account
+	// any defaults applied by Postgres, and expressed as a space separated string of
+	// ROLE PRIVILEGE clauses.
+	PrivilegesAsClauses string `json:"privilegesAsClauses,omitempty"`
```

Comment:
For some reason I was under the impression that Kubernetes' `printcolumn` functionality would not handle arrays properly - outputting as a string was just the default because that is what is passed directly into the query to create / update privileges.

I missed adding any output columns to `Grant` so I've added these as well, and since the privileges there are represented as an array I've changed the output of both to be 'array-like' for consistency:

```
NAME           READY   SYNCED   CONN LIMIT   PRIVILEGES
example-role   True    True     -1           [CREATEDB INHERIT NOBYPASSRLS NOCREATEROLE NOLOGIN NOREPLICATION NOSUPERUSER]
parent-role    True    True     10           [INHERIT LOGIN NOBYPASSRLS NOCREATEDB NOCREATEROLE NOREPLICATION NOSUPERUSER]

...

NAME                               READY   SYNCED   AGE   ROLE           MEMBER OF     DATABASE   PRIVILEGES
example-grant-role-1-on-database   True    True     12m   example-role                 example    [CREATE]
example-grant-role-2-on-database   True    True     12m   example-role                 example    [CONNECT TEMPORARY]
example-grant-role-membership      True    True     12m   example-role   parent-role
```